### PR TITLE
statistics: add `undersized-region-count` and `oversized-region-count` unhealthy region type 

### LINKED ISSUE #4783: add monitor to record some unhealthy region such as`need-merge` and `need-split` region.
## Enhancement Task
To know the count of the unhealthy region if some configuration about region changed, pd should add some monitor to record some unhealthy region such asneed-merge and need-split region.

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?

Issue Number: Close #4783

statistics: add `undersized-region-count` and `oversized-region-count` unhealthy region type 

### What is changed and how it works?

```commit-message
```

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/statistics/region_collection.go [hunk 1]
```diff
@@ -163,6 +172,10 @@ func (r *RegionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
 		PendingPeer: len(region.GetPendingPeers()) > 0,
 		LearnerPeer: len(region.GetLearners()) > 0,
 		EmptyRegion: region.GetApproximateSize() <= core.EmptyRegionApproximateSize,
+		SplitRegion: region.GetApproximateSize() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxSize()) ||
+			region.GetApproximateKeys() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxKeys()),
+		MergerRegion: region.GetApproximateSize() < int64(r.opt.GetScheduleConfig().MaxMergeRegionSize) &&
+			region.GetApproximateSize() < int64(r.opt.GetScheduleConfig().MaxMergeRegionKeys),
```
Comment (rleungx) [inline server/statistics/region_collection.go:178]: It only takes the size/key into account, but there might be lots of factors that decide if a region needs to be merged. So the metrics of the output of `pd-ctl` may cause misleading.

Comment (JmPotato) [reply server/statistics/region_collection.go:178]: What about changing the name from `SplitRegion` and `MergerRegion` into `OversizeRegion` and `UndersizeRegion`? This should be more proper to the actual meaning.

File: server/statistics/region_collection.go [hunk 2]
```diff
@@ -240,6 +255,8 @@ func (r *RegionStatistics) Collect() {
 	offlineRegionStatusGauge.WithLabelValues("learner-peer-region-count").Set(float64(len(r.offlineStats[LearnerPeer])))
 	offlineRegionStatusGauge.WithLabelValues("empty-region-count").Set(float64(len(r.offlineStats[EmptyRegion])))
 	offlineRegionStatusGauge.WithLabelValues("offline-peer-region-count").Set(float64(len(r.offlineStats[OfflinePeer])))
+	offlineRegionStatusGauge.WithLabelValues("over-size-region-count").Set(float64(len(r.offlineStats[OversizeRegion])))
+	offlineRegionStatusGauge.WithLabelValues("under-size-region-count").Set(float64(len(r.offlineStats[UndersizeRegion])))
```
Comment (rleungx) [inline server/statistics/region_collection.go:259]: Maybe we don't need them.

File: server/statistics/region_collection.go [hunk 3]
```diff
@@ -232,6 +245,8 @@ func (r *RegionStatistics) Collect() {
 	regionStatusGauge.WithLabelValues("pending-peer-region-count").Set(float64(len(r.stats[PendingPeer])))
 	regionStatusGauge.WithLabelValues("learner-peer-region-count").Set(float64(len(r.stats[LearnerPeer])))
 	regionStatusGauge.WithLabelValues("empty-region-count").Set(float64(len(r.stats[EmptyRegion])))
+	regionStatusGauge.WithLabelValues("over-size-region-count").Set(float64(len(r.stats[OversizeRegion])))
+	regionStatusGauge.WithLabelValues("under-size-region-count").Set(float64(len(r.stats[UndersizeRegion])))
```
Comment (rleungx) [inline server/statistics/region_collection.go:249]: It seems all regions will be divided into these two types. So maybe we can use another monitor to record these normal situations.

File: server/api/region.go [hunk 4]
```diff
@@ -498,6 +498,40 @@ func (h *regionsHandler) GetOfflinePeerRegions(w http.ResponseWriter, r *http.Re
 	h.rd.JSON(w, http.StatusOK, regionsInfo)
 }
 
+// @Tags region
+// @Summary List all regions that needs merges.
```
Comment (JmPotato) [inline server/api/region.go:502]:
```suggestion
// @Summary List all regions that are oversized.
```

File: server/api/region.go [hunk 5]
```diff
@@ -498,6 +498,40 @@ func (h *regionsHandler) GetOfflinePeerRegions(w http.ResponseWriter, r *http.Re
...
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	regionsInfo := convertToAPIRegions(regions)
+	h.rd.JSON(w, http.StatusOK, regionsInfo)
+}
+
+// @Tags region
+// @Summary List all regions that needs split.
```
Comment (JmPotato) [inline server/api/region.go:519]:
```suggestion
// @Summary List all regions that are undersized.
```

File: server/api/region.go [hunk 6]
```diff
@@ -498,6 +498,40 @@ func (h *regionsHandler) GetOfflinePeerRegions(w http.ResponseWriter, r *http.Re
 	h.rd.JSON(w, http.StatusOK, regionsInfo)
 }
 
+// @Tags region
+// @Summary List all regions that needs merges.
+// @Produce json
+// @Success 200 {object} RegionsInfo
+// @Failure 500 {string} string "PD server failed to proceed the request."
+// @Router /regions/check/oversize-region [get]
+func (h *regionsHandler) GetOverSizeRegions(w http.ResponseWriter, r *http.Request) {
```
Comment (JmPotato) [inline server/api/region.go:507]:
```suggestion
func (h *regionsHandler) GetOversizedRegions(w http.ResponseWriter, r *http.Request) {
```

File: server/api/region.go [hunk 7]
```diff
@@ -498,6 +498,40 @@ func (h *regionsHandler) GetOfflinePeerRegions(w http.ResponseWriter, r *http.Re
...
+	h.rd.JSON(w, http.StatusOK, regionsInfo)
+}
+
+// @Tags region
+// @Summary List all regions that needs split.
+// @Produce json
+// @Success 200 {object} RegionsInfo
+// @Failure 500 {string} string "PD server failed to proceed the request."
+// @Router /regions/check/undersize-region [get]
+func (h *regionsHandler) GetUndersizeRegions(w http.ResponseWriter, r *http.Request) {
```
Comment (JmPotato) [inline server/api/region.go:524]:
```suggestion
func (h *regionsHandler) GetUndersizedRegions(w http.ResponseWriter, r *http.Request) {
```

File: tests/pdctl/region/region_test.go [hunk 8]
```diff
@@ -144,6 +144,12 @@ func (s *regionTestSuite) TestRegion(c *C) {
 		{[]string{"region", "check", "down-peer"}, []*core.RegionInfo{r3}},
 		// region check learner-peer command
 		{[]string{"region", "check", "learner-peer"}, []*core.RegionInfo{r3}},
+		// region check empty-region command
+		{[]string{"region", "check", "empty-region"}, []*core.RegionInfo{r1}},
+		// region check merge-region command
+		{[]string{"region", "check", "undersize-region"}, []*core.RegionInfo{r1, r4}},
+		// region check split-region command
```
Comment (JmPotato) [inline tests/pdctl/region/region_test.go:151]: The comments need to be updated also.

**Review by JmPotato — COMMENTED** [review]
Please rename the title and description of this PR also.

**Review by AndreMouche — APPROVED** [review]
LGTM

File: server/api/region.go [hunk 9]
```diff
@@ -498,6 +498,40 @@ func (h *regionsHandler) GetOfflinePeerRegions(w http.ResponseWriter, r *http.Re
...
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	regionsInfo := convertToAPIRegions(regions)
+	h.rd.JSON(w, http.StatusOK, regionsInfo)
+}
+
+// @Tags region
+// @Summary List all regions that are under merge sized.
```
Comment (AndreMouche) [inline server/api/region.go:519]: `under merge sized.` makes me confused, how about `are undersized`  directly?

Comment (bufferflies) [reply server/api/region.go:519]: yes , it will be better to describe 
.

File: server/statistics/region_collection.go [hunk 10]
```diff
@@ -35,6 +35,8 @@ const (
 	OfflinePeer
 	LearnerPeer
 	EmptyRegion
+	OversizeRegion
+	UndersizeRegion
```
Comment (JmPotato) [inline server/statistics/region_collection.go:39]:
```suggestion
	OversizedRegion
	UndersizedRegion
```

File: server/statistics/region_collection.go [hunk 11]
```diff
@@ -71,15 +76,15 @@ func NewRegionStatistics(opt *config.PersistOptions, ruleManager *placement.Rule
 	r.stats[PendingPeer] = make(map[uint64]*RegionInfo)
 	r.stats[LearnerPeer] = make(map[uint64]*RegionInfo)
 	r.stats[EmptyRegion] = make(map[uint64]*RegionInfo)
+	r.stats[OversizeRegion] = make(map[uint64]*RegionInfo)
+	r.stats[UndersizeRegion] = make(map[uint64]*RegionInfo)
```
Comment (JmPotato) [inline server/statistics/region_collection.go:80]:
```suggestion
	r.stats[OversizedRegion] = make(map[uint64]*RegionInfo)
	r.stats[UndersizedRegion] = make(map[uint64]*RegionInfo)
```

File: server/statistics/region_collection.go [hunk 12]
```diff
@@ -163,11 +168,15 @@ func (r *RegionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
 		PendingPeer: len(region.GetPendingPeers()) > 0,
 		LearnerPeer: len(region.GetLearners()) > 0,
 		EmptyRegion: region.GetApproximateSize() <= core.EmptyRegionApproximateSize,
+		OversizeRegion: region.GetApproximateSize() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxSize()) ||
+			region.GetApproximateKeys() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxKeys()),
+		UndersizeRegion: region.GetApproximateSize() < int64(r.opt.GetScheduleConfig().MaxMergeRegionSize) &&
```
Comment (JmPotato) [inline server/statistics/region_collection.go:173]:
```suggestion
		OversizedRegion: region.GetApproximateSize() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxSize()) ||
			region.GetApproximateKeys() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxKeys()),
		UndersizedRegion: region.GetApproximateSize() < int64(r.opt.GetScheduleConfig().MaxMergeRegionSize) &&
```

File: server/statistics/region_collection.go [hunk 13]
```diff
@@ -232,13 +241,14 @@ func (r *RegionStatistics) Collect() {
 	regionStatusGauge.WithLabelValues("pending-peer-region-count").Set(float64(len(r.stats[PendingPeer])))
 	regionStatusGauge.WithLabelValues("learner-peer-region-count").Set(float64(len(r.stats[LearnerPeer])))
 	regionStatusGauge.WithLabelValues("empty-region-count").Set(float64(len(r.stats[EmptyRegion])))
+	regionStatusGauge.WithLabelValues("over-size-region-count").Set(float64(len(r.stats[OversizeRegion])))
+	regionStatusGauge.WithLabelValues("under-size-region-count").Set(float64(len(r.stats[UndersizeRegion])))
```
Comment (JmPotato) [inline server/statistics/region_collection.go:245]:
```suggestion
	regionStatusGauge.WithLabelValues("oversized-region-count").Set(float64(len(r.stats[OversizedRegion])))
	regionStatusGauge.WithLabelValues("undersized-region-count").Set(float64(len(r.stats[UndersizedRegion])))
```

Comment (bufferflies) [reply server/statistics/region_collection.go]: fix

File: server/statistics/region_collection.go [hunk 14]
```diff
@@ -163,11 +168,15 @@ func (r *RegionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
...
+		OversizedRegion: region.GetApproximateSize() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxSize()) ||
+			region.GetApproximateKeys() >= int64(r.storeManager.GetStoreConfig().GetRegionMaxKeys()),
+		UndersizedRegion: region.GetApproximateSize() < int64(r.opt.GetScheduleConfig().MaxMergeRegionSize) &&
+			region.GetApproximateSize() < int64(r.opt.GetScheduleConfig().MaxMergeRegionKeys),
 	}
 
 	for typ, c := range conditions {
 		if c {
-			if isRemoving {
+			if isRemoving && typ < EmptyRegion {
```
Comment (JmPotato) [inline server/statistics/region_collection.go:179]: Should be `<=` here?

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: server/statistics/region_collection.go (hunk 14 above)
Comment (bufferflies) [reply server/statistics/region_collection.go:179]: It should use `<`, the offlineStats is used to analyze the process of the offline store,  so the empty regions are not belong them.

**Review by JmPotato — APPROVED** [review]

**Pull request merged**